org/mozilla/jss/provider/java/security/RSAPSSAlgorithmParameters.java

```diff
@@ -0,0 +1,244 @@
+            throw new IOException("Invalid encoded data.");
+        }
+
+        DerInputStream mgf1Str = new DerInputStream(seq[1].toByteArray());
+        DerValue[] seqMgf1 = mgf1Str.getSequence(2);
+
+        ObjectIdentifier mgf1OID = seqMgf1[0].getOID();
+
+        if (!mgf1OID.equals(AlgorithmId.MGF1_oid)) {
+           throw new IOException("Invalid encoded data: expected MGF1 OID but got: " + mgf1OID.toString());
+        } else {
+           specMGF1Name = "MGF1";
+        }
+
+        if (seq[2].isContextSpecific((byte)2)) {
+            seq[2]  = seq[2].data.getDerValue();
+        } else {
+            throw new IOException("Invalid encoded data.");
+        }
+
+        BigInt sLength = seq[2].getInteger();
+
+        this.spec = new PSSParameterSpec(specAlgName, specMGF1Name, specMFG1ParamSpec,
+                                         sLength.toInt(), 1 /* always default trailer */);
+
+        populateFromSpec();
+    }
+
+    private void encode(DerOutputStream out) throws IOException {
+        try (
+            DerOutputStream tmp = new DerOutputStream();
+            DerOutputStream mgf = new DerOutputStream();
+            DerOutputStream seq1 = new DerOutputStream();
+            DerOutputStream intStream = new DerOutputStream();
+        ) {
+            // Hash algorithm
+            hashAlg.derEncodeWithContext(tmp,0);
+
+            // Mask Gen Function Sequence
+            mgf.putOID(maskGenFunc.getOID());
+
+            // MGF hash alg is the same as the hash Alg at this point.
+            hashAlg.encode(mgf);
+            seq1.write(DerValue.tag_Sequence,mgf);
+            tmp.write(DerValue.createTag(DerValue.TAG_CONTEXT,
+                                             true, (byte) 1), seq1);
+
+            // Salt Length
+            intStream.putInteger(saltLen);
+
+            tmp.write(DerValue.createTag(DerValue.TAG_CONTEXT, true, (byte) 2),
+                    intStream);
+
+            // Ignore trailer field, it never changes over all sequence tags
+            out.write(DerValue.tag_Sequence, tmp);
+
+            byte[] data = out.toByteArray();
+        }
+    }
+
+    private void populateFromSpec() {
+        if (spec == null || hashAlg == null) {
+            return;
+        }
+
+        String hashAlgName = spec.getDigestAlgorithm();
+        String maskGenName = spec.getMGFAlgorithm();
+
+        int saltLen = spec.getSaltLength();
+        this.saltLen = new BigInt(saltLen);
+        int trailer = spec.getTrailerField();
+
+        // Create the hash alg and mask gen func objects
+        if (hashAlgName.equals("SHA-256")) {
+           hashAlg = new AlgorithmId(AlgorithmId.SHA256_oid);
+        }  else if(hashAlgName.equals("SHA-512")) {
+           hashAlg = new AlgorithmId(AlgorithmId.SHA512_oid);
+        }  else if(hashAlgName.equals("SHA-384")) {
+           hashAlg = new AlgorithmId(AlgorithmId.SHA384_oid);
+        } else {
+           hashAlg = new AlgorithmId(AlgorithmId.SHA_oid);
```

Comment:
@jmagne Why do we fallback to SHA1 here? Is there a relevant portion of the spec somewhere?

Reply:
I just put that there cause other implementations of this thing I observed that being the default value. I can see throwing an exception could be appropriate here..